src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/MultiSourcePackageMetadataProvider.cs

```diff
@@ -292,6 +303,19 @@ private static IEnumerable<PackageVulnerabilityMetadata> MergeVulnerabilityMetad
             return vulnerabilityMetadatas.FirstOrDefault(v => v != null && v.Any());
         }
 
+        private static string MergePackagePath(IEnumerable<IPackageSearchMetadata> packages)
+        {
+            var cpPaths = packages
+                .OfType<ClonedPackageSearchMetadata>()
+                .Where(packageMetadata => !string.IsNullOrWhiteSpace(packageMetadata.PackagePath))
+                .Select(packageMetadata => packageMetadata.PackagePath);
+            var lpPaths = packages
+                .OfType<LocalPackageSearchMetadata>()
+                .Where(packageMetadata => !string.IsNullOrWhiteSpace(packageMetadata.PackagePath))
+                .Select(packageMetadata => packageMetadata.PackagePath);
+            return cpPaths.Concat(lpPaths).FirstOrDefault();
```

Comment:
I don't know how other people feel about LINQ in non perf-critical code paths, but this is going to cause at minimum 5 enumerators to be allocated on the heap, if the first non-null package path is found in a `ClonedPackageSearchMetadata`, thanks to LINQ's lazy evaluation. However, if there are no packages with a non-null `PackagePath` of type ClonedPackageSearchMetadata, then there will be an additional 3 enumerators allocated on the heap, a total of 8.

If this is replaced with a foreach loop over the `packages` enumerable, then only a single alloc is needed, and honestly I think that using a switch expression, it could be the same number of lines of code.

Having written all that, searching for `MergePackagePath` doesn't find any usage of this method.

Reply:
I was using it back when I was doing the readme extraction in the client code. Removed